### ISSUE TITLE
Redundant VRouter guest network on wrong interface

### ISSUE THREAD

**Issue by DennisKonrad** [apache/cloudstack#3179]
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
redundant VPC Offering
~~~

##### CLOUDSTACK VERSION
~~~
master
~~~

##### CONFIGURATION
Standard cluster deployment regarding this issue. No changes made to sytemvm or the offerings involved.

##### OS / ENVIRONMENT
Seems to work only with OvS bridges.
KVM with OvS
Projects with VPCs

##### SUMMARY
Redundant vRouter Offering doesn't work because guest network is on wrong interface in vRouter

##### STEPS TO REPRODUCE

- Add VPC; choose Redundant VPC Offering
- Create a guest network/instance to allow for redundancy to build up
- some network is created on the wrong interface:

![redundantrouter](https://user-images.githubusercontent.com/33484272/52776520-0d434d00-3042-11e9-8863-50c005b26847.PNG)

![image](https://user-images.githubusercontent.com/33484272/68468805-0cc69400-0219-11ea-810d-2f83643e7c30.png)

##### EXPECTED RESULTS

The guest network should be added on eth1 because this is the interface where both vRouters share the MAC address.

##### ACTUAL RESULTS
Redundancy state of the redundant routers changes from UNKNOWN to FAILED because conntrackd and keepalived cannot establish the VRRP redundancy between the two vRouters.

**Comment by svenvogel** [comment]
@fmaximus can you help to debug this case? you helped already with this pr #2304 by vpc. maybe you can help with redudant vpc. i hope. thx

**Comment by fmaximus** [comment]
@DennisKonrad Can you post the contents of /etc/cloudstack/ips.json on the VR.

**Comment by DennisKonrad** [comment]
@fmaximus Thank you for helping. This is the content of the ips.json:

{
  "eth0": [
    {
      "add": true,
      "broadcast": "169.254.255.255",
      "cidr": "169.254.2.133/16",
      "device": "eth0",
      "gateway": "",
      "netmask": "255.255.0.0",
      "network": "169.254.0.0/16",
      "nic_dev_id": "0",
      "nw_type": "control",
      "one_to_one_nat": false,
      "public_ip": "169.254.2.133",
      "size": "16",
      "source_nat": false
    }
  ],
  "eth2": [
    {
      "add": true,
      "broadcast": "xxx.xxx.xxx.xxx",
      "cidr": "xxx.xxx.xxx.xxx/26",
      "device": "eth2",
      "first_i_p": true,
      "gateway": "xxx.xxx.xxx.1",
      "netmask": "255.255.255.192",
      "network": "xxx.xxx.xxx.0/26",
      "new_nic": true,
      "nic_dev_id": 2,
      "nw_type": "public",
      "one_to_one_nat": false,
      "public_ip": "xxx.xxx.xxx.56",
      "size": "26",
      "source_nat": true,
      "vif_mac_address": "1e:00:93:00:00:9a"
    },
    {
      "add": true,
      "broadcast": "10.32.64.255",
      "cidr": "10.32.64.45/24",
      "device": "eth2",
      "gateway": "10.32.64.1",
      "netmask": "255.255.255.0",
      "network": "10.32.64.0/24",
      "nic_dev_id": "2",
      "nw_type": "guest",
      "one_to_one_nat": false,
      "public_ip": "10.32.64.45",
      "size": "24",
      "source_nat": false
    }
  ],
  "id": "ips"
}

**Comment by svenvogel** [comment]
@fmaximus Hi Frank, thank you for help. do you need additional informations?

**Comment by ustcweizhou** [comment]
@svenvogel Are you using kvm ?

**Comment by DennisKonrad** [comment]
@ustcweizhou Yes we do

**Comment by fmaximus** [comment]
@DennisKonrad What I can see in the ips.json, is that mac address is missing.
The lookup uses the mac address to find out the device id.

**Comment by svenvogel** [comment]
@fmaximus Yes you are right. do you know where in the code is the problem so why redundant router is missing some mac adresses?

**Comment by DennisKonrad** [comment]
@fmaximus @svenvogel 
Hi ,
does this mean when the code that's responsible for updating the "ips.json" would also insert the mac address this could fix the problem?

I searched for the ips.json in the code but wasn't able to locate the code responsible for it. 
Any hints are much appreciated

**Comment by ustcweizhou** [comment]
@DennisKonrad 
could you provide more information ?
(1) which cloudstack version ?
(2) Do you use official systemvm template ?

some other comments
(1) eth1 should be public ip, and vpc tier ip will be attached to eth2/eth3, etc
(2) ips are saved into ips.json in systemvm/debian/opt/cloud/bin/merge.py
(3) restart VPC with cleanup might fix your problem.

**Comment by DennisKonrad** [comment]
@ustcweizhou 
info
(1) we are running more or less the master on the agents and mgmts 
     The systemvm template is built from master too but from december 2018
(2) There are no modifications done to the systemvm template

other comments
(1) yes, eth1->public, eth2++ -> guest networks, VRRP->lowest guest nic
**_!! Do you think the problem is, that the public ip is on the wrong interface when using redundant vrouters? (And not the guest ips as I thought...) _**

(2) I tried to read through this but got lost. It's really hard to see from where some functions are supposed to be called. Is there someone who knows this code?
(3) I will try to do that today

**Comment by ustcweizhou** [comment]
@DennisKonrad 
when apply a new configuration in VR, it will create a file in VR and execute the update in VR.
see code in ./core/src/com/cloud/agent/resource/virtualnetwork/VirtualRoutingResource.java below
```
    private ExecutionResult applyConfigToVR(String routerAccessIp, ConfigItem c, Duration timeout) {
        if (c instanceof FileConfigItem) {
            FileConfigItem configItem = (FileConfigItem)c;
            return _vrDeployer.createFileInVR(routerAccessIp, configItem.getFilePath(), configItem.getFileName(), configItem.getFileContents());
        } else if (c instanceof ScriptConfigItem) {
            ScriptConfigItem configItem = (ScriptConfigItem)c;
            return _vrDeployer.executeInVR(routerAccessIp, configItem.getScript(), configItem.getArgs(), timeout);
        }
        throw new CloudRuntimeException("Unable to apply unknown configitem of type " + c.getClass().getSimpleName());
    }
```

json files will be created in server, then scp to VR through 169.254.X.X/port 3922 on VR.
then execute /opt/cloud/bin/update_config.py in VR
json file will be processed in following part
```
def process_file():
    logging.info("Processing JSON file %s" % sys.argv[1])
    qf = QueueFile()
    qf.setFile(sys.argv[1])
    qf.load(None)
```
QueueFile is defined in /opt/cloud/bin/merge.py
ips.json will be updated if json file contains new ip.

**Comment by svenvogel** [comment]
@fmaximus does the info from @ustcweizhou help for a deeper look from you? maybe you have the time.

**Comment by DennisKonrad** [comment]
I found some more time looking into this. @ustcweizhou Do you also think the problem originates in 
https://github.com/apache/cloudstack/blob/52f68a273aaf0aa6746eaaee4a95a7bf722c53b0/core/src/main/java/com/cloud/agent/resource/virtualnetwork/VirtualRoutingResource.java#L401

**Comment by DennisKonrad** [comment]
The root cause seems to be that the interfaces (eth1, eth2, eth3) share the same mac address.
I'm not really sure where this originates. Somehow when the redundant router is created and the nics are added all of them get the same mac.
The "createRedundantAssociateIPCommands" seems to assign the ips to the wrong interface then.

@ustcweizhou Can you give me a hint where in the code new vrouters get their nics?

**Comment by ustcweizhou** [comment]
@DennisKonrad I tried to reproduce the issue today but failed with cloudstack 4.11.2

**Comment by DennisKonrad** [comment]
@ustcweizhou 
Are you using KVM with OVS? We are using advanced networking.
Maybe we can pin this down by identifying the difference.

**Comment by ustcweizhou** [comment]
@DennisKonrad KVM with linux bridge. We do not use OVS.

**Comment by rhtyd** [comment]
@DennisKonrad can you confirm if you managed to fix the issue? Are you using OVS with KVM? I could not reproduce the error on KVM/centos7 with standard Linux-bridge networking.

**Comment by rhtyd** [comment]
This is a known issue and it's likely that we are unable to fix issue and the PR seems to be adding regressions. I've moved it to 4.13.1.0 milestone for now, if the regression/issues can be fixed we can consider including the PR towards 4.13.0.0.

**Comment by ustcweizhou** [comment]
@DennisKonrad 
Does the issue happen occasionally ? Can it be reproduced by the steps you mentioned?

**Comment by DennisKonrad** [comment]
Everytime we try to deploy redundant offering for VPCs.

It only seems to happen with OvS bridges though. I tried to reproduce it with Linux Bridges but there it did work.

**Comment by ustcweizhou** [comment]
@DennisKonrad have you tried with offering with single router?

**Comment by DennisKonrad** [comment]
@ustcweizhou The single Router offering works like a charm in thus setup. We use that for hundreds of VMs

**Comment by weizhouapache** [comment]
@DennisKonrad now we can conclude the issue happens with ovs and redundant VPC offering.
did you have this issue in old versions ? for example, 4.9.X or 4.11.X ?

**Comment by DennisKonrad** [comment]
@ustcweizhou It's very likely the issue is very old. My colleague mentioned it could be as old as 4.6 ...

We are currently trying to find out more where the issue originates. It looks like the hint fmaximus gave is in parts true:
In the json file the VPC-router receives, the dev_num and vif_mac address are wrong for the public ip. So whereever this gets matched the error happens.

**Comment by weizhouapache** [comment]
@DennisKonrad can you reproduce the issue and upload some mgt logs ?
you can grep the log by job id (job-XXXXX).

**Comment by DennisKonrad** [comment]
@weizhouapache The management logs are mostly uninteresting. But they also show that both VPC-router go into BACKUP state:

![image](https://user-images.githubusercontent.com/33484272/68592698-f5dfa600-0493-11ea-86e3-319135eb5237.png)
![image](https://user-images.githubusercontent.com/33484272/68592737-0db72a00-0494-11ea-81ac-d756cbbdf6cc.png)

**Comment by DennisKonrad** [comment]
Also we checked and found out, that the redundant VPC ist not using the https://github.com/apache/cloudstack/blob/9e6ee7c63e8b3073f9855511fe7cb004414b2dd4/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L1756

We think that this is very likely the root of this issue.

**Comment by weizhouapache** [comment]
@DennisKonrad we have a fix for another issue with VPC, maybe it solves this bug.

```
diff --git a/server/src/main/java/com/cloud/network/router/CommandSetupHelper.java b/server/src/main/java/com/cloud/network/router/CommandSetupHelper.java
index f9ffeb9..e2da241 100644
--- a/server/src/main/java/com/cloud/network/router/CommandSetupHelper.java
+++ b/server/src/main/java/com/cloud/network/router/CommandSetupHelper.java
@@ -866,7 +866,12 @@ public class CommandSetupHelper {
                 }
             }

-            final IpAssocCommand cmd = new IpAssocCommand(ipsToSend);
+            final IpAssocCommand cmd;
+            if (router.getVpcId() != null) {
+                cmd = new IpAssocVpcCommand(ipsToSend);
+            } else {
+                cmd = new IpAssocCommand(ipsToSend);
+            }
             cmd.setAccessDetail(NetworkElementCommand.ROUTER_IP, _routerControlHelper.getRouterControlIp(router.getId()));
             cmd.setAccessDetail(NetworkElementCommand.ROUTER_GUEST_IP, _routerControlHelper.getRouterIpInNetwork(associatedWithNetworkId, router.getId()));
             cmd.setAccessDetail(NetworkElementCommand.ROUTER_NAME, router.getInstanceName());
```

can you try with it ?

**Comment by DennisKonrad** [comment]
@weizhouapache Thanks! Looks very promising. I will compile/test and come back to report if that fixes it.

**Comment by DennisKonrad** [comment]
@weizhouapache I tested now and hit a related error. In https://github.com/apache/cloudstack/blob/96d98de85cdce3ab9e33c6dda6ffa131d825cd85/systemvm/debian/opt/cloud/bin/cs/CsAddress.py#L261
it fails with:
_self.dnum = hex(100 + int(dev[3:]))ValueError: invalid literal for int() with base 10: 'None'_

After that the deployment of the redundant vpc fails and the router is deleted (The second one does not even start to deploy). I'll try to step through code tomorrow and see why there is no device number. Also I will try to get my hands on to the generated json file.

Do you see anything suspicious?

**Comment by weizhouapache** [comment]
@DennisKonrad 
I will test it as well

**Comment by DennisKonrad** [comment]
@weizhouapache I found that the json the router receives really does not have the "nic_dev_id" set at all. Therefor the script fails because the value has "null" asigned then.

**redundant VPC**
`{"ip_address":[
{"public_ip":"172.20.20.12","source_nat":true,"add":true,"one_to_one_nat":false,"first_i_p":true,"gateway":"172.20.0.1","netmask":"255.255.0.0","vif_mac_address":"1e:00:d5:00:00:c2","new_nic":false,"nw_type":"Public"}],"type":"ips"}

**non redundant VPC**
{"ip_address":[
{"public_ip":"172.20.20.13","source_nat":true,"add":true,"one_to_one_nat":false,"first_i_p":true,"gateway":"172.20.0.1","netmask":"255.255.0.0","vif_mac_address":"1e:00:f4:00:00:c3",**"nic_dev_id":1**,"new_nic":false,"nw_type":"Public"}],"type":"ips"}`

**Comment by weizhouapache** [comment]
@DennisKonrad 
I reverted #2304, and it works

**Comment by weizhouapache** [comment]
@DennisKonrad 
I believe this is a regress issue caused by #2304
I think we should revert #2304, and look for another solution for the issue mentioned in in the title of #2304
However, I did not find more information about the issue.

**Comment by svenvogel** [comment]
@weizhouapache Hi Weizhou, sorry that I rummage in between. i think #2304 was introduced from @fmaximus because we had some problem with the interface order on "OVS", "OpenvSwitch". There was no possibility to add a nat or something like that. Now linux bridges and OVS will work. From my side a better way would be to change the redundant VR component like the non redundant. 

Can you support on this side? Can you give us some hints how and where we change the redundant part of VR?

**Comment by weizhouapache** [comment]
@svenvogel @DennisKonrad 

there are two methods to handle ip association in LibvirtComputingResource.java
```
public ExecutionResult prepareNetworkElementCommand(final IpAssocCommand cmd) {
protected ExecutionResult prepareNetworkElementCommand(final IpAssocVpcCommand cmd) {
```
In the past they have almost same processing. @fmaximus has changed the second method in #2304 for openvswitch setup. the first method is not changed which might be the root cause of this issue #3679 @DennisKonrad .

For now all ip association in normal setup(vpc with rvr, linux bridge) are handled by the first method, which work fine.  I did not see any IpAssocVpcCommand (all are IpAssocCommand ) so I am confused how issue mentioned in #2304 is fixed. @svenvogel can you give me more details ?

I found the mac address of nic and ip addresses are different, which might be why nic device id is determined by broadcast uri instead of mac address, and also the reason device id is null reported by @DennisKonrad  in https://github.com/apache/cloudstack/issues/3179#issuecomment-553940664

in my opinion, we should 
(1) apply the change in https://github.com/apache/cloudstack/issues/3179#issuecomment-552463825
(2) different process for ovs and linux bridge in the two method above. to be clear, nic device id is determined by mac address map (if ovs) or broadcast url (if linux bridge).

**Comment by DennisKonrad** [comment]
@weizhouapache The second function 
`prepareNetworkElementCommand(final IpAssocVpcCommand cmd)`
 does work for linux bridge and ovs at the moment I think. But it doen't work for the redundant vpc offering with ovs.

Can you explain why you want to
> (1) apply the change in #3179 (comment)

Do you think it would be possible to write a mac address map based function for both (ovs and linux bridge)?

**Comment by weizhouapache** [comment]
@DennisKonrad 
I did not see any IpAssocVpcCommand  in management server log, so I think all are IpAssocCommand , even when ip is associated to a VPC. Please correct me if I am wrong

We have seen and fixed many issues with VPC & RvRs in cloudstack 4.7.1, especially when there are multiple public ip ranges.
For example, network element should not be cleaned up when de-associate a public ip.
https://github.com/apache/cloudstack/blob/master/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L377-L382
If  IP is associated/de-associated to an isolated network, the command should be IpAssocCommand 
If  IP is associated/de-associated to a VPC, the command should be IpAssocVpcCommand 

the second function
```prepareNetworkElementCommand(final IpAssocVpcCommand cmd)```
does NOT work for normal setup (linux bridge), if we apply the change in https://github.com/apache/cloudstack/issues/3179#issuecomment-552463825
as you said in https://github.com/apache/cloudstack/issues/3179#issuecomment-553641686

I have changed the first function to similar behavior as the second function. I got the same result as the description. Now I can conclude that the issue is a regression issue after #2304, but the root cause should be wrong mac address of public ip addresses.
https://github.com/apache/cloudstack/blob/master/server/src/main/java/com/cloud/network/router/CommandSetupHelper.java#L812-L814

**Comment by weizhouapache** [comment]
@DennisKonrad @svenvogel 
I will create a PR when I test ok with VPC & RvR & linux bridge.

**Comment by DennisKonrad** [comment]
@weizhouapache Yes you are right:

> the second function
prepareNetworkElementCommand(final IpAssocVpcCommand cmd)
does NOT work for normal setup (linux bridge), if we apply the change in #3179 (comment)

That's the reason why I asked about applying the change in https://github.com/apache/cloudstack/issues/3179.

We also thought about another bug beeing the root cause. A wrong MAC seems very plausible. I will help test the PR.

**Comment by weizhouapache** [comment]
@DennisKonrad 
if you look at CommandSetupHelper.java, you may find that IpAssocVpcCommand is send to hypervisors when associate a public IP, if VPC does not have redundant VRs. However, for VPC with RvRs, it always send IpAssocCommand (in createRedundantAssociateIPCommands), which does not make any sense.

There should be some issues caused by it in the past, some of them have been fixed by different way, for example 6fc389a044288e8f12f41a651a711a1ab7113b86 . However, I insist that we should send IpAssocVpcCommand instead of IpAssocCommand , even it might lead to some issues.

**Comment by weizhouapache** [comment]
@DennisKonrad 
I have set up a testing environment with OVS. I cannot reproduce this issue.
Do you test on official cloudstack release or with some your changes ?

**Comment by svenvogel** [comment]
@ustcweizhou thanks for the test. We encounter the problem in 2016 that VPC was working from outside but if we use a NAT with Public IP it was not working. Thats because Public Adress was not set on correct interface (dev id). this was the fixed from @fmaximus for the non-redundant router.

two questions:

You are sure thats working in your test environment with ovs? (NAT, Firewall ...) if its so then we need to test it without #2304

did you remove #2304 in your ovs test setup and used the change from #3179?

can you explain how this commit help 6fc389a in our actually problem?

did you test both redundant and non-redundant router?

>>However, I insist that we should send IpAssocVpcCommand instead of IpAssocCommand , even it might lead to some issues.

like @DennisKonrad wrote above. if we use "IpAssocVpcCommand" we get the issues with the VPC redundant router. they are start up and reboot immediately. that why he ask  "Do you think it would be possible to write a mac address map based function for both (ovs and linux bridge)?" thats why we ask can you fix this problem? 

```json
`{"ip_address":[
{"public_ip":"172.20.20.12","source_nat":true,"add":true,"one_to_one_nat":false,"first_i_p":true,"gateway":"172.20.0.1","netmask":"255.255.0.0","vif_mac_address":"1e:00:d5:00:00:c2","new_nic":false,"nw_type":"Public"}],"type":"ips"}```

**Comment by weizhouapache** [comment]
@svenvogel #2304 is helpful for VPC with single router. we do not need to revert it.
we need to make similar change for OVS in LibvirtComputingResource.java. I have found some issues with isolated networks in my setup with OVS.

**Comment by DennisKonrad** [comment]
@weizhouapache Leaving https://github.com/apache/cloudstack/pull/2304 in place sounds good. To me it looks like the more stable approach.

Do you think the changes to get the isolated networks up to speed require some heavy lifting? If theres something to test I can do so.

**Comment by DennisKonrad** [comment]
@weizhouapache I'll setup a test environment to test the ovs part. It will take some days. I'll come back with feedback

**Comment by DennisKonrad** [comment]
@weizhouapache Just writing a short feedback so you know where I'm at. At the moment it seems the redundant VPC does not deploy with your changes.

After deploying cloudstack management isn't able to connect to the agent running on the first VPC router. (The second one never get's started because the first one fails.) My guess is that the APIPA address is the culprit.

I will setup a systemvm with password access so I can look into the VPC router and give you more information on that.
The non-redundant VPC deploys just fine.

**Comment by DennisKonrad** [comment]
@weizhouapache After a first look into the router I can say the APIPA is not the problem. There's something else wrong. I will have a look and come back to report

**Comment by weizhouapache** [comment]
@DennisKonrad it is still in progress. The issue you mentioned should be caused by the last commit.
It is difficult to test all scenarios. I will write some test scripts at first.

**Comment by DennisKonrad** [comment]
@weizhouapache Ok, kindly give me a ping if there's something to test or help.

**Comment by weizhouapache** [comment]
@DennisKonrad I have uploaded a commit to fix the regression issue.
https://github.com/ustcweizhou/cloudstack/commits/4.14.0-vpc-vr-issues
I am writing scripts for automation test.

**Comment by DennisKonrad** [comment]
@weizhouapache The first test looks promising. The issue with the agent is gone. Redundant routers are deploying fine now. Also it looks like the VRRP IP is on the right interface. The routers are negotiating BACKUP and Master. 

I'll go on and test NAT and features. I'll report back when done.

**Comment by DennisKonrad** [comment]
@weizhouapache I tested multiple things that work fine (for redundant VPC offering):

- [x] NAT is working (one NAT ip)
- [x] Settings seem to go to both routers
- [x] Redundancy heals when one router dies
- [x] Failover of traffic works when one master dies
- [x] Adding multiple Guest Networks works

For things that do not work I have found:

At the moment it's only possible to deploy one VM in a GuestNet (when using redundant VPC with OvS). For the second VM an error occurs. If one deletes that one working VM it's possible to deploy again. But max 1 VM. Even if I try to deploy in another GuestNet (in the same VPC).

Management log only shows:
<details><summary>click to unfold</summary>
<p>

```INFO  [o.a.c.a.c.u.v.DeployVMCmd] (API-Job-Executor-34:ctx-6242853f job-107 ctx-b55ef55c) (logid:852387c4) Unable to create a deployment for VM[User|i-4-18-VM]
com.cloud.exception.InsufficientServerCapacityException: Unable to create a deployment for VM[User|i-4-18-VM]Scope=interface com.cloud.dc.DataCenter; id=1
	at org.apache.cloudstack.engine.cloud.entity.api.VMEntityManagerImpl.reserveVirtualMachine(VMEntityManagerImpl.java:215)
	at org.apache.cloudstack.engine.cloud.entity.api.VirtualMachineEntityImpl.reserve(VirtualMachineEntityImpl.java:200)
	at com.cloud.vm.UserVmManagerImpl.startVirtualMachine(UserVmManagerImpl.java:4676)
	at com.cloud.vm.UserVmManagerImpl.startVirtualMachine(UserVmManagerImpl.java:4231)
	at com.cloud.vm.UserVmManagerImpl.startVirtualMachine(UserVmManagerImpl.java:4221)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:338)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:197)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:174)
	at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:51)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:174)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy204.startVirtualMachine(Unknown Source)
	at org.apache.cloudstack.api.command.user.vm.DeployVMCmd.execute(DeployVMCmd.java:588)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:156)
	at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:108)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:583)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:531)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
</p>
</details>

<br>
Also I noticed another error in the management log that could be related:

<details><summary>click to unfold</summary>
<p>

```ERROR [o.a.c.e.o.NetworkOrchestrator] (API-Job-Executor-32:ctx-d8346e02 job-92 ctx-5a62bb49) (logid:c681419d) Failed to remove dhcp-dns entry due to: 
com.cloud.exception.ResourceUnavailableException: Resource [DataCenter:1] is unreachable: Can't find at least one router!
	at com.cloud.network.element.VirtualRouterElement.removeDhcpEntry(VirtualRouterElement.java:961)
	at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.cleanupNicDhcpDnsEntry(NetworkOrchestrator.java:3029)
	at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.cleanupNics(NetworkOrchestrator.java:2023)
	at com.cloud.vm.VirtualMachineManagerImpl.advanceExpunge(VirtualMachineManagerImpl.java:525)
	at com.cloud.vm.VirtualMachineManagerImpl.advanceExpunge(VirtualMachineManagerImpl.java:490)
	at com.cloud.vm.UserVmManagerImpl.expunge(UserVmManagerImpl.java:2191)
	at com.cloud.vm.UserVmManagerImpl.destroyVm(UserVmManagerImpl.java:2876)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:338)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:197)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:174)
	at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:51)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:174)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy204.destroyVm(Unknown Source)
	at org.apache.cloudstack.api.command.user.vm.DestroyVMCmd.execute(DestroyVMCmd.java:137)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:156)
	at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:108)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:583)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:49)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:46)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:531)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
</p>
</details>

**Comment by weizhouapache** [comment]
@DennisKonrad 
thanks for feedback !

InsufficientServerCapacityException might be caused by two reasons
(1) capacity on the platform
(2) some issues when apply network rules in virtual routers. Can you see some Exception in the agent.log on the server where the VR is running ?

**Comment by DennisKonrad** [comment]
@ustcweizhou You are completely right. There was a ressource issue i failed to see. With that gone multiple VMs deploy just fine.

I will start testing multiple public IPs and address ranges

**Comment by weizhouapache** [comment]
@DennisKonrad 
vpc looks good now. However I found an issue with isolated networks (rvr or non-rvr both)

**Comment by weizhouapache** [comment]
@DennisKonrad 
do you use the network offering created by following http://docs.cloudstack.apache.org/en/latest/plugins/ovs-plugin.html ?

**Comment by DennisKonrad** [comment]
@weizhouapache Hi, no we don't use the ovs plugin. We have OvS in place for the hypervisor bridges.

I'll resume testing today. I would like to test if maybe the make_redundant process offered for non-redundant VPC is working. Also I'll test if custom network offerings are posing problems

**Comment by DennisKonrad** [comment]
@weizhouapache The "make redundant" worked fine for me. I also looked if private gateways go to the right interface and as far as I can tell this looks good.

Can you tell me something about the issue with isolated networks so maybe I can replicate the error and give you some feedback?

**Comment by svenvogel** [comment]
@weizhouapache If you are ready is it possible to create an PR with the changes? 👍

**Comment by weizhouapache** [comment]
@svenvogel there are still some issues in my testing.
I will create a PR when everything is fully tested.

**Comment by weizhouapache** [comment]
@svenvogel @DennisKonrad 
I push 3 new commits to the branch.

**Comment by DennisKonrad** [comment]
@weizhouapache I'll build and test it for OvS today. Anything you want me to test?

**Comment by DaanHoogland** [comment]
there is a lot of text here. Do we have plans for 4.13 @weizhouapache @DennisKonrad @svenvogel ?

**Comment by weizhouapache** [comment]
@DaanHoogland what's the plan for 4.13.1 and 4.14 ?
sorry I came back from holiday today and missed some discussion.

**Comment by svenvogel** [comment]
@ustcweizhou Hi ... back from holiday. i dont know whats tested. there are things from your side they needs to be tested so that we come ready to a PR?

**Comment by svenvogel** [comment]
@weizhouapache ping 😄

**Comment by weizhouapache** [comment]
@svenvogel I will create a PR next week.
I did not fully test it. Hope it will be good :-D

**Comment by svenvogel** [comment]
@weizhouapache i think so 😁 we will test it again its maybe easier but what i saw was very good!

**Comment by DennisKonrad** [comment]
@weizhouapache Hi, sorry for the long delay. The last time I tested this I couldn't find any problems. The two setups I tested (KVM with Linux Bridge and KVM with OVS) looked very good

**Comment by DaanHoogland** [comment]
@DennisKonrad i read in your text above we can close this?

**Comment by DennisKonrad** [comment]
@DaanHoogland I would like to wait for the PR of wei to be tested and merged. If the community accepts it then we can close the issue.

**Comment by donnydavis** [comment]
This is still an issue for me. Non redundant routers work fine, redundant routers do not.

**Comment by weizhouapache** [comment]
@DennisKonrad @svenvogel @DaanHoogland 
I am testing my changes. If everything goes fine I will create pr tomorrow.

**Comment by svenvogel** [comment]
@weizhouapache that sounds great. Big thanks! 👍

**Comment by DaanHoogland** [comment]
msolution merged with PR #3847
